File: implementations/go/sentinelblocks_test.go
package sentinelblocks

import "testing"

const fenced = "<<<JSON>>>\n```json\n{\"name\": \"Ada\", \"age\": 36,}\n```\n<<<END>>>"
const multi = "<<<FILE src/a.ts>>>\nexport const a = 1;\n<<<END>>>\n" +
	"<<<FILE src/b.ts>>>\nexport const b = 2;\n<<<END>>>"

func TestSentinelFenceTrailingComma(t *testing.T) {
	v, err := JSONFromResponse(fenced, "")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	m, ok := v.(map[string]interface{})
	if !ok || m["name"] != "Ada" || m["age"].(float64) != 36 {
		t.Fatalf("got %#v", v)
	}
}

func TestBareJSONFallback(t *testing.T) {
	v, err := JSONFromResponse(`{"x": 10}`, "")
	if err != nil || v.(map[string]interface{})["x"].(float64) != 10 {
		t.Fatalf("got %#v err %v", v, err)
	}
}

func TestBraceInStringBalanced(t *testing.T) {
	got, ok := FirstJSONObject(`prefix {"k":"a}b{c"} suffix`)
	if !ok || got != `{"k":"a}b{c"}` {
		t.Fatalf("got %q ok %v", got, ok)
	}
}

func TestNamedMultiBlock(t *testing.T) {
	b := ExtractBlocks(multi, "FILE")
	if len(b) != 2 || b[0] != "export const a = 1;" || b[1] != "export const b = 2;" {
		t.Fatalf("got %#v", b)
	}
}

func TestTaggedArg(t *testing.T) {
	tagged := ExtractTaggedBlocks(multi, "FILE")
	if len(tagged) != 2 || tagged[0].Arg != "src/a.ts" || tagged[1].Arg != "src/b.ts" {
		t.Fatalf("got %#v", tagged)
	}
}

func TestFileIsNotFiles(t *testing.T) {
	if _, ok := ExtractBlock("<<<FILES>>>\nnope\n<<<END>>>", "FILE"); ok {
		t.Fatal("FILE should not match FILES")
	}
}

func TestErrorOnNoJSON(t *testing.T) {
	if _, err := JSONFromResponse("no json here at all", ""); err == nil {
		t.Fatal("expected error")
	}
}

func TestWrapRoundTrip(t *testing.T) {
	got, ok := ExtractBlock(Wrap("JSON", `{"ok":true}`), "JSON")
	if !ok || got != `{"ok":true}` {
		t.Fatalf("got %q ok %v", got, ok)
	}
	tagged := ExtractTaggedBlocks(WrapNamed("FILE", "src/x.go", "code"), "FILE")
	if len(tagged) != 1 || tagged[0].Arg != "src/x.go" {
		t.Fatalf("got %#v", tagged)
	}
}

File: implementations/java/SentinelBlocksTest.java
// Invariant suite — Java. Build:
//   javac SentinelBlocks.java SentinelBlocksTest.java && java SentinelBlocksTest
import java.util.List;

public class SentinelBlocksTest {
    static int pass = 0, fail = 0;

    static void check(String name, boolean cond) {
        if (cond) { pass++; System.out.println("  ok  - " + name); }
        else      { fail++; System.out.println("  FAIL- " + name); }
    }

    public static void main(String[] args) {
        String fenced = "<<<JSON>>>\n```json\n{\"name\": \"Ada\", \"age\": 36,}\n```\n<<<END>>>";
        String multi =
            "<<<FILE src/a.ts>>>\nexport const a = 1;\n<<<END>>>\n" +
            "<<<FILE src/b.ts>>>\nexport const b = 2;\n<<<END>>>";

        check("sentinel+fence+trailing-comma",
              SentinelBlocks.jsonTextFromResponse(fenced, null)
                  .equals(java.util.Optional.of("{\"name\": \"Ada\", \"age\": 36}")));

        check("bare-json fallback",
              SentinelBlocks.jsonTextFromResponse("{\"x\": 10}", null)
                  .equals(java.util.Optional.of("{\"x\": 10}")));

        check("brace-in-string balanced",
              SentinelBlocks.firstJsonObject("prefix {\"k\":\"a}b{c\"} suffix")
                  .equals(java.util.Optional.of("{\"k\":\"a}b{c\"}")));

        List<String> blocks = SentinelBlocks.extractBlocks(multi, "FILE");
        check("named multi-block count", blocks.size() == 2);
        check("named multi-block content",
              blocks.size() == 2 && blocks.get(0).equals("export const a = 1;")
              && blocks.get(1).equals("export const b = 2;"));

        var tagged = SentinelBlocks.extractTaggedBlocks(multi, "FILE");
        check("tagged arg",
              tagged.size() == 2 && tagged.get(0).arg().equals("src/a.ts")
              && tagged.get(1).arg().equals("src/b.ts"));

        check("FILE != FILES",
              SentinelBlocks.extractBlock("<<<FILES>>>\nnope\n<<<END>>>", "FILE").isEmpty());

        check("none on no-json",
              SentinelBlocks.jsonTextFromResponse("no json here at all", null).isEmpty());

        String w = SentinelBlocks.wrap("JSON", "{\"ok\":true}");
        check("wrap round-trip",
              SentinelBlocks.extractBlock(w, "JSON").equals(java.util.Optional.of("{\"ok\":true}")));

        System.out.println("\nJava: " + pass + " passed, " + fail + " failed");
        if (fail > 0) System.exit(1);
    }
}
